tornado/wsgi.py

```diff
@@ -89,11 +93,33 @@ async def main():
     The `tornado.web.FallbackHandler` class is often useful for mixing
     Tornado and WSGI apps in the same server.  See
     https://github.com/bdarnell/django-tornado-demo for a complete example.
+
+    `WSGIContainer` supports executing the WSGI application in custom executors
+    using `IOLoop.run_in_executor`. The default executor uses
+    `tornado.concurrent.dummy_executor` which works synchronously, but other
+    executors subclassing `concurrent.futures.Executor` may be used. To execute
+    WSGI application code in separate threads in an event-loop compatible way
+    use::
+
+        async def main():
+            executor = concurrent.futures.ThreadPoolExecutor()
+            # ^-- Execute requests in separate threads.
+            container = tornado.wsgi.WSGIContainer(simple_app, executor)
+            # Initialize the WSGI container with custom executor --^
+            http_server = tornado.httpserver.HTTPServer(container)
+            http_server.listen(8888)
+            await asyncio.Event().wait()
+
+    Running the WSGI app with a `ThreadPoolExecutor` remains *less scalable*
+    than running the same app in a multi-threaded WSGI server like ``gunicorn``
+    or ``uwsgi``.
     """
 
-    def __init__(self, wsgi_application: "WSGIAppType") -> None:
+    def __init__(self, wsgi_application: "WSGIAppType", executor: futures.Executor = None) -> None:
         self.wsgi_application = wsgi_application
+        self.executor = concurrent.dummy_executor if executor is None else executor
 
+    @gen.coroutine
```

Comment:
We nearly always want to use `async def` instead of `@gen.coroutine` now. I suspect that you may have done this because the caller isn't expecting a coroutine and `gen.coroutines` start themselves in this case while native coroutines require something to either await them or explicitly start them. 

It's probably better to use a native coroutine and explicitly start it than to use the older interface. Maybe even refactor this to use the newer delegate interfaces, although I'm not sure those add much value unless we start implementing some sort of streaming for request bodies.

Reply:
I wasn't able to get it working as an `async def` and I couldn't figure out why. In the updated PR, I see you do `IOLoop.current().spawn_callback(self.handle_request, request)`. If that works, it's fine but it's rather magical to me. Maybe worth a comment.